src/renderer/routes/test_sequencer_panel/components/DataTable.tsx

```diff
@@ -0,0 +1,525 @@
+  );
+  const [columnVisibility, setColumnVisibility] =
+    React.useState<VisibilityState>({});
+  const [rowSelection, setRowSelection] = React.useState({});
+  const data = elems; // this is necessary for some reason for the table to work no idea why
+
+  const table = useReactTable({
+    data,
+    columns,
+    onSortingChange: setSorting,
+    onColumnFiltersChange: setColumnFilters,
+    getCoreRowModel: getCoreRowModel(),
+    pageCount: 5,
+    getPaginationRowModel: getPaginationRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+    onColumnVisibilityChange: setColumnVisibility,
+    onRowSelectionChange: setRowSelection,
+    state: {
+      sorting,
+      columnFilters,
+      columnVisibility,
+      rowSelection,
+    },
+  });
+
+  const handleClickRemoveTests = () => {
+    onRemoveTest(map(Object.keys(rowSelection), (idxStr) => parseInt(idxStr)));
+    setRowSelection([]);
+  };
+
+  const onRemoveTest = (idxs: number[]) => {
+    setElems((elems) => {
+      //first, collect all idxs to remove in a set
+      const toRemove = new Set();
+      idxs.forEach((idx) => {
+        toRemove.add(elems[idx].groupId);
+      });
+      return filter(elems, (elem) => !toRemove.has(elem.groupId));
+    });
+  };
+
+  const [showAddConditionalModal, setShowAddConditionalModal] =
+    React.useState(false);
+  const addConditionalAfterIdx = React.useRef(-1);
+
+  const [showWriteConditionalModal, setShowWriteConditionalModal] =
+    React.useState(false);
+  const writeConditionalForIdx = React.useRef(-1);
+
+  const handleWriteConditionalModal = (input: string) => {
+    setElems((data) => {
+      const new_data = [...data];
+      const conditional = new_data[
+        writeConditionalForIdx.current
+      ] as Conditional;
+      new_data[writeConditionalForIdx.current] = {
+        ...conditional,
+        condition: input,
+      };
+      return new_data;
+    });
+  };
+
+  const handleAddConditionalModal = (type: ConditionalComponent) => {
+    setElems((data) => {
+      const new_data = [...data];
+      new_data.splice(
+        addConditionalAfterIdx.current,
+        0,
+        ...generateConditional(type),
+      );
+      return new_data;
+    });
+  };
+
+  const handleClickAddConditional = (idx: number) => {
+    addConditionalAfterIdx.current = idx;
+    setShowAddConditionalModal(true);
+  };
+
+  const onClickWriteCondition = (idx: number) => {
+    writeConditionalForIdx.current = idx;
+    setShowWriteConditionalModal(true);
+  };
+
+  const getSpecificContextMenuItems = (row: Row<TestSequenceElement>) => {
+    switch (row.original.type) {
+      case "test":
+        return <></>;
+      case "conditional":
+        return (
+          <>
+            <ContextMenuItem
+              onClick={() => onClickWriteCondition(parseInt(row.id))}
+            >
+              Write Condition
+            </ContextMenuItem>
+          </>
+        );
+    }
+  };
```

Comment:
you don't need fragments here, just return null in the first case and return the lone element in the second

Reply:
this will probably change in the future since we will need to add some context menu items for tests. Keeping it this way allows to more easily add these features once we need them.